(PDB-1252) Remove clamq; use ActiveMQ directly

rbrw opened the pull request:
(PDB-1252) Remove clamq; use ActiveMQ directly
    
Remove clamq entirely, and instead, just call ActiveMQ and javax.jms
directly.

pljenkinsro commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.puppetlabs.com/job/platform_puppetdb_intn-sys_pr/682/

grimradical commented:
This is good stuff, and long overdue.

Minor point: I think in a few spots you refer to JMX (management/metrics extensions) when I think you mean JMS (messaging). :)

rbrw commented:
@grimradical Oh, right -- fixed.  And thanks.

pljenkinsro commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.puppetlabs.com/job/platform_puppetdb_intn-sys_pr/800/

pljenkinsro commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.puppetlabs.com/job/platform_puppetdb_intn-sys_pr/817/

rbrw commented:
@pljenkinsro retest this please

pljenkinsro commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.puppetlabs.com/job/platform_puppetdb_intn-sys_pr/826/

wkalt commented:
@pljenkinsro retest this please

pljenkinsro commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.puppetlabs.com/job/platform_puppetdb_intn-sys_pr/838/

pljenkinsro commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.puppetlabs.com/job/platform_puppetdb_intn-sys_pr/907/

pljenkinsro commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.puppetlabs.com/job/platform_puppetdb_intn-sys_pr/941/

mullr commented on project.clj:
```diff
@@ -38,7 +38,7 @@
                  [metrics-clojure "0.7.0" :exclusions [org.clojure/clojure org.slf4j/slf4j-api]]
                  [clj-time "0.5.1"]
-                 [org.clojure/java.jmx "0.2.0"]
+                 [org.clojure/java.jmx "0.3.0"]
```
0.3.1 was just released a few days ago, we should probably use that.

mullr commented on project.clj:
```diff
@@ -55,10 +55,14 @@
+                 [org.apache.activemq/activemq-pool "5.7.0"
+                  :exclusions [org.slf4j/slf4j-api]]
+
                  ;; bridge to allow some spring/activemq stuff to log over slf4j
```
Is the 'spring' part of this still accurate?

rbrw replied:
@mullr They were when I made these changes, but I'll be checking again before the final version.

rbrw replied:
@mullr ahh, so I thought you were talking about the exclusions, which I did double-check that we need.  With respect to the comment, I suppose it's accurate in the sense that we do still pull in some spring bits via activemq.

mullr commented on src/puppetlabs/puppetdb/command.clj:
```diff
@@ -131,27 +132,29 @@
           (string? mq-endpoint)
           (string? raw-command)]
    :post [(string? %)]}
+  ;; FIXME: is this efficient enough?  It's rebuilding the whole MQ
```
This is what we talked about before, right? I'd love to see this fixed, but I'd be totally fine with making it a separate ticket.o

rbrw replied:
@mullr Already fixed here -- I've reworked all of that side to share a single connection.

wkalt replied:
comment needs to be removed at any rate

mullr commented on src/puppetlabs/puppetdb/command.clj:
```diff
@@ -131,27 +132,29 @@
     id))
 
-(defn enqueue-command!
+(defn-validated enqueue-command! :- s/Str
```
Question: when do we use defn-validated in this codebase? What's the motivation for using it here? Don't read too much in to this, I just want to know.

pljenkinsro commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.puppetlabs.com/job/platform_puppetdb_intn-sys_pr/991/

pljenkinsro commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.puppetlabs.com/job/platform_puppetdb_intn-sys_pr/993/

pljenkinsro commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.puppetlabs.com/job/platform_puppetdb_intn-sys_pr/995/

pljenkinsro commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.puppetlabs.com/job/platform_puppetdb_intn-sys_pr/994/

pljenkinsro commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.puppetlabs.com/job/platform_puppetdb_intn-sys_pr/996/

rbrw commented on src/puppetlabs/puppetdb/cli/services.clj:
```diff
@@ -204,30 +202,43 @@
   [context]
+  ;; FIXME: wait for in-flight jetty calls, and then block access
+  ;; here, say via wrap-with-puppetdb-middleware returning 503 or
+  ;; similar? (preexisting condition)
```
I've filed a separate ticket for this (PDB-1388).

rbrw commented:
@pljenkinsro retest this please

pljenkinsro commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.puppetlabs.com/job/platform_puppetdb_intn-sys_pr/997/

rbrw commented:
@mullr @senior hopefully I've addressed your comments, and I've squashed all the commits.  The only remaining issues are the three "Q:" comments regarding exception handling.

pljenkinsro commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.puppetlabs.com/job/platform_puppetdb_intn-sys_pr/999/

rbrw commented:
@pljenkinsro retest this please

rbrw commented:
Note that there was some additional discussion here:  https://github.com/rbrw/puppetdb/commit/ec57e88b2aea86e09b226d976a733b1a515c8d9e#commitcomment-10566369

pljenkinsro commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.puppetlabs.com/job/platform_puppetdb_intn-sys_pr/1000/

wkalt commented on src/puppetlabs/puppetdb/cli/services.clj:
```diff
@@ -78,15 +79,10 @@
-  [command version payload]
-  (command/enqueue-command! mq-addr mq-endpoint command version payload))
-
 (defn auto-deactivate-nodes!
```
this is being called with a partial below as if it still only had one argument. I think that's what's causing the test failues.

wkalt replied:
looks like you can just change it to #(auto-deactivate-nodes! node-ttl % mq-connection) and everything will work.

rbrw replied:
@wkalt of course, thanks for tracking that down.

wkalt commented on src/puppetlabs/puppetdb/command.clj:
```diff
@@ -119,40 +120,30 @@
+(defn-validated enqueue-raw-command! :- s/Str
+  "Submits raw-command to the mq-endpoint of mq-connection and returns
+  its id. Annotates the command via annotate-command, unless it can't
+  be parsed."
```
does this actually do anything with annotation?

rbrw replied:
@wkalt ahh, good catch.

wkalt commented on src/puppetlabs/puppetdb/mq.clj:
```diff
@@ -142,81 +136,107 @@
-      (throw @mq-error)
-      (vec (take limit @contents)))))
+;; In the final patch, perhaps move the code here, but this
+;; should make review easier.
```
marking for visibility

rbrw commented on src/puppetlabs/puppetdb/cli/services.clj:
```diff
@@ -329,8 +332,10 @@
+        ;; Run database maintenance tasks seqentially to avoid
+        ;; competition. Each task must handle its own errors.
+        (gc-task #(doseq [maintain (remove nil? db-maintenance-tasks)]
+                    (maintain write-db)))
```
@wkalt does this look OK to you, or think I should revert it?  (I thought it was a bit more straightforward.)

wkalt replied:
@rbrw I agree, this seems nicer to me.

pljenkinsro commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.puppetlabs.com/job/platform_puppetdb_intn-sys_pr/1003/

rbrw commented:
@pljenkinsro retest this please

pljenkinsro commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.puppetlabs.com/job/platform_puppetdb_intn-sys_pr/1004/

wkalt commented:
This causes an ugly JMS error to get logged (to the log, not the console) every time puppetdb is stopped. You can reproduce by tailing the log and killing PDB with kill or ctrl-c.

wkalt commented on src/puppetlabs/puppetdb/mq_listener.clj:
```diff
@@ -393,31 +331,72 @@
+           (mq/commit-or-rollback sess (handle (mq/convert-jms-message msg)))
+           (catch Throwable ex
+             ;; Q: What do we really want to do here?
+             (log/error ex "message receive failed")
```
I think all the Q: comments can be removed without code changes, unless there's something you think should be done in these cases aside from logging the exception.

rbrw replied:
@wkalt done, thanks.

pljenkinsro commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.puppetlabs.com/job/platform_puppetdb_intn-sys_pr/1008/

wkalt commented:
I'm +1 after tests pass; looks like @mullr's comment about spring also needs resolution.

wkalt commented:
@pljenkinsro retest this please

mullr commented:
:+1: from me

pljenkinsro commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.puppetlabs.com/job/platform_puppetdb_intn-sys_pr/1009/
